File: ShortestDistanceII.java
// Time: O(L+K) | Space: O(N)

class WordDistance {

    Map<String, List<Integer>> map;
    public WordDistance(String[] wordsDict) {
        map = new HashMap<>();
        for(int i=0;i<wordsDict.length;i++) {
            if(!map.containsKey(wordsDict[i]))
                map.put(wordsDict[i], new ArrayList<>());
            map.get(wordsDict[i]).add(i);
        }
    }

    public int shortest(String word1, String word2) {
        List<Integer> w1l= map.get(word1);
        List<Integer> w2l = map.get(word2);
        int i=0;
        int j=0;
        int min = Integer.MAX_VALUE;
        while(i<w1l.size() && j<w2l.size()) {
            int firstIdx = w1l.get(i);
            int secondIdx = w2l.get(j);
            min = Math.min(min, Math.abs(firstIdx-secondIdx));
            if(firstIdx < secondIdx) {
                i++;
            } else {
                j++;
            }
        }
        return min;
    }
}

/**
 * Your WordDistance object will be instantiated and called as such:
 * WordDistance obj = new WordDistance(wordsDict);
 * int param_1 = obj.shortest(word1,word2);
 */

File: shortestDistanceII.cpp
// Time complexity: O(m+n) where m is the size of indices in 1 and n in indices2 
// Space complexity: O(N) number of words

// Approach : hashmap with 2 pointer to get the result
/*
Hashmap solution is more suitable since we can use the constructor to create the hashmap 

2 pointers to get the final solution

*/

class WordDistance {
public:
    unordered_map<string, vector<int>> dict;
    WordDistance(vector<string>& wordsDict) {
        for(int i = 0; i<wordsDict.size(); i++) {
            dict[wordsDict[i]].push_back(i);
        }
    }
    
    int shortest(string word1, string word2) {
        vector<int> i1 = dict[word1];
        vector<int> i2 = dict[word2];

        int a = 0, b = 0;
        int result = INT_MAX;
        while(a < i1.size() && b < i2.size()) {
            result = min(result, abs(i1[a] - i2[b]));
            if(i1[a] < i2[b])
                a++;
            else
                b++;
        }
        return result;
    }
};

/**
 * Your WordDistance object will be instantiated and called as such:
 * WordDistance* obj = new WordDistance(wordsDict);
 * int param_1 = obj->shortest(word1,word2);
 */